PSReadLine/History.cs

```diff
@@ -137,15 +137,27 @@ private AddToHistoryOption GetAddToHistoryOption(string line)
                 }
 
                 object value = Options.AddToHistoryHandler(line);
+                if (value is PSObject psObj)
+                {
+                    value = psObj.BaseObject;
+                }
+
+                if (value is bool boolValue)
+                {
+                    return boolValue ? AddToHistoryOption.MemoryAndFile : AddToHistoryOption.SkipAdding;
+                }
 
-                if (LanguagePrimitives.TryConvertTo(value, out AddToHistoryOption enumValue))
+                if (value is AddToHistoryOption enumValue)
                 {
                     return enumValue;
                 }
 
-                if (value is bool boolValue && !boolValue)
+                // 'TryConvertTo' incurs exception handling when the value cannot be converted to the target type.
```

Comment:
I'll bet some people will return strings that convert to `AddToHistoryOption`. You could make those conversions cheaper with a tiny bit of special case conversion code here.

Reply:
Agreed. The special case for string conversion is added.